lib/logout.js

```diff
@@ -0,0 +1,363 @@
+        });
+
+        // Update reference to include signature
+        options.reference =  constants.ELEMENTS.LOGOUT_RESPONSE.SIGNATURE_LOCATION_PATH;
+        options.destination = data.serviceProviderLogoutURL || options.destination;
+        // We stored the relay state of the initial request
+        options.relayState = transaction.relayState;
+        prepareAndSendToken(req, res, 'LOGOUT_RESPONSE', logoutResponse, options, next);
+      });
+    });
+  }
+
+  function validateSamlResponse(req, sessions, cb) {
+    var SAMLResponse = req.query.SAMLResponse || req.body.SAMLResponse;
+
+    function parseAndValidate(err, buffer) {
+      if (err) { return cb(err); }
+      var xml = new DOMParser().parseFromString(buffer.toString());
+      var parsedResponse = {};
+
+      // status code
+      var statusCodes = xml.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:protocol', 'StatusCode');
+      var statusCodeXml = statusCodes[0];
+      if (statusCodeXml) {
+        parsedResponse.status = statusCodeXml.getAttribute('Value');
+
+        // status sub code
+        var statusSubCodeXml = statusCodes[1];
+        if (statusSubCodeXml) {
+          parsedResponse.subCode = statusSubCodeXml.getAttribute('Value');
+        }
+      }
+
+      // status message
+      var samlStatusMsgXml = xml.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:protocol', 'StatusMessage')[0];
+      if (samlStatusMsgXml) {
+        parsedResponse.message = samlStatusMsgXml.textContent;
+      }
+
+      // status detail
+      var samlStatusDetailXml = xml.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:protocol', 'StatusDetail')[0];
+      if (samlStatusDetailXml) {
+        parsedResponse.detail = samlStatusDetailXml.textContent;
+      }
+      // Issuer
+      var issuer = xml.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:assertion', 'Issuer')[0];
+      if (issuer) {
+        parsedResponse.issuer = issuer.textContent;
+      }
+
+      req.parsedSAMLResponse = parsedResponse;
+      var sessionParticipants = getSession(sessions, parsedResponse.issuer);
+      
+      if (!sessionParticipants.cert){ 
+        // If there's no certificate to check the signature, let it pass
+        return cb(null, parsedResponse);
+      }
+
+      // validate signature
+      try {
+        var validationOptions = xtend({signingCert: sessionParticipants.cert}, options);
+        utils.validateSignature(req, 'RESPONSE', xml, validationOptions);
+      } catch (e) {
+        return cb(e);
+      }
+
+      cb(null, parsedResponse);
+    }
+
+    if (req.body.SAMLResponse || !options.deflate) {
+      // HTTP-POST or HTTP-Redirect without deflate encoding
+      return parseAndValidate(null, new Buffer(SAMLResponse, 'base64'));
+    }
+
+    // Default: HTTP-Redirect with deflate encoding
+    zlib.inflateRaw(new Buffer(SAMLResponse, 'base64'), parseAndValidate);
+  }
+
+  return function (req, res, next) {
+    try {
```

Comment:
There are a lot of things that could go wrong, specially with signing and signature verification. We choose to stay safe using this general try/catch approach